openmls/src/group/managed_group/validation.rs

```diff
@@ -0,0 +1,244 @@
+use mls_group::{proposals::StagedProposal, staged_commit::StagedCommit};
+use tls_codec::Serialize;
+
+use super::*;
+
+impl ManagedGroup {
+    /// This function is used to parse messages from the DS.
+    /// It checks for syntactic errors and makes some semantic checks as well.
+    /// If the input is a [MlsCiphertext] message, it will be decrypted.
+    /// Returns an [UnverifiedMessage] that can be inspected and later processed in
+    /// [self::process_unverified_message()].
+    pub fn parse_message(
+        &mut self,
+        message: InboundMessage,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<UnverifiedMessage, ManagedGroupError> {
+        /*
+        High level checks:
+         - epoch must be within bounds
+         - Sender must be a member
+         - AAD can be extracted/evaluated
+         - decryption
+         - IFF content_type is a commit, confirmation_tag must be present
+         - membership tag must be present/verified
+        */
+
+        // Make sure we are still a member of the group
+        if !self.active {
+            return Err(ManagedGroupError::UseAfterEviction(UseAfterEviction::Error));
+        }
+        // Check the message has the correct epoch number
+        if message.epoch() != self.group.context().epoch() {
+            return Err(ManagedGroupError::InvalidMessage(
+                InvalidMessageError::WrongEpoch,
+            ));
+        }
+
+        // Check the type of message we received
+        let mut membership_tag_required = false;
+        let (plaintext, aad) = match message {
+            // If it is a ciphertext we decrypt it and return the plaintext message.
+            // Attempting to decrypt it will also check the bounds of the epoch.
+            InboundMessage::Ciphertext(ciphertext) => {
+                let aad = ciphertext.authenticated_data.clone().into_vec();
+
+                let plaintext = self.group.decrypt(&ciphertext, backend)?;
+
+                (plaintext, aad)
+            }
+            // If it is a plaintext message we return it with an empty AAD
```

Comment:
Wouldn't `Option` be more suitable for the AAD here? Otherwise we can't distinguish between an empty AAD and a non-existent one.